src/lightning/fabric/utilities/spike.py

```diff
@@ -0,0 +1,165 @@
+
+    Args:
+        mode: Whether to minimize or maximize the tracked metric
+        window: A running mean of metrics with ``window`` size. Serves as reference value for spike3s
+        warmup: After how many batches should spike-tracking starts
+        atol: An absolute tolerance.  Every diff between the running mean and the current value,
+            that's not an improvement and above ``atol`` will be considered a spike
+        rtol: A relative tolerance. Every diff between the running mean and the current value,
+            that's higher than ``rtol * running_mean`` is considered a spike
+        exclude_batches_path: Where to save the file that contains the batches to exclude. Will default to current directory.
+        finite_only: Will consider non-finite values like NaN, inf and -inf a spike as well
+    """
+
+    def __init__(
+        self,
+        mode: Literal["min", "max"] = "min",
+        window: int = 10,
+        warmup: int = 1,
+        atol: Optional[float] = None,
+        rtol: Optional[float] = 2.0,
+        exclude_batches_path: Optional[str] = None,
+        finite_only: bool = True,
+    ):
+        if _TORCHMETRICS_GREATER_EQUAL_1_0_0:
+            from torchmetrics.aggregation import MeanMetric
+            from torchmetrics.wrappers import Running
+        else:
+            raise RuntimeError("SpikeDetection requires torchmetrics>=1.0.0! Please upgrade your version!")
+        super().__init__()
+
+        self.last_val: Union[torch.Tensor, float] = 0.0
+        # spike detection happens individually on each machine
+        self.running_mean = Running(MeanMetric(dist_sync_on_step=False, sync_on_compute=False), window=window)
+        self.mode = mode
+        self.warmup = warmup
+        self.atol = atol
+        self.rtol = rtol
+        self.bad_batches: List[int] = []
+        self.exclude_batches_path = exclude_batches_path
+        self.finite_only = finite_only
+
+    @torch.no_grad()
+    def on_train_batch_end(self, fabric: "Fabric", loss: torch.Tensor, batch: Any, batch_idx: int) -> None:
+        """Checks if we currently have a loss-spike."""
+        if batch_idx == 0:
+            self.running_mean.to(fabric.strategy.root_device)
+
+        if self.exclude_batches_path is None:
+            self.exclude_batches_path = os.getcwd()
+
+        if not str(self.exclude_batches_path).endswith(".json"):
+            self.exclude_batches_path = os.path.join(self.exclude_batches_path, "skip_batches.json")
+
+        is_spike = bool(batch_idx >= self.warmup and self.is_spike(loss))
+
+        # While spike-detection happens on a per-rank level, we need to fail all ranks if any rank detected a spike
+        is_spike_global = fabric.strategy.reduce_boolean_decision(is_spike, all=False)
+
+        if is_spike_global:
+            self.handle_spike(fabric, batch_idx)
+        else:
+            self.update_stats(loss)
+
+    def is_spike(self, loss: torch.Tensor) -> bool:
```

Comment:
I would make all methods protected except the hooks to call (on_train_batch_end) 2) and the state_dict methods. To clearly communicate what the user is supposed to call manually. I consider this especially important for Fabric where the user needs to call the hook.

Reply:
done :)